Data/HashMap/Internal.hs

```diff
@@ -2456,3 +2465,90 @@ instance (Eq k, Hashable k) => Exts.IsList (HashMap k v) where
     type Item (HashMap k v) = (k, v)
     fromList = fromList
     toList   = toList
+
+------------------------------------------------------------------------
+-- Debugging
+
+data Validity k = Invalid (Error k) SubHashPath | Valid
+  deriving (Eq, Show)
+
+instance Semigroup (Validity k) where
+  Valid <> y = y
+  x     <> _ = x
+
+instance Monoid (Validity k) where
+  mempty = Valid
+  mappend = (<>)
+
+data Error k
+  = INV1_internal_Empty
+  | INV2_misplaced_hash !Hash
+  | INV3_key_hash_mismatch k !Hash
+  | INV4_Collision_size !Int
+  | INV5_Collision_duplicate_key k !Hash
+  | INV6_bad_BitmapIndexed_size !Int
+  | INV7_bitmap_array_size_mismatch !Bitmap !Int
+  | INV8_BitmapIndexed_invalid_single_subtree
+  | INV9_bad_Full_size !Int
+  deriving (Eq, Show)
+
+-- | A part of a 'Hash' with 'bitsPerSubkey' bits.
+type SubHash = Word
+
+data SubHashPath = Root | Cons !SubHash !SubHashPath
+  deriving (Eq, Show)
+
+valid :: Hashable k => HashMap k v -> Validity k
+valid Empty = Valid
+valid t     = validInternal Root t
+  where
+    validInternal p Empty                 = Invalid INV1_internal_Empty p
+    validInternal p (Leaf h l)            = validHash p h <> validLeaf p h l
+    validInternal p (Collision h ary)     = validHash p h <> validCollision p h ary
+    validInternal p (BitmapIndexed b ary) = validBitmapIndexed p b ary
+    validInternal p (Full ary)            = validFull p ary
+
+    validHash p0 h0 = go p0 h0
+      where
+        go Root !_ = Valid
+        go (Cons sh p) h | h .&. subkeyMask == sh = go p (h `unsafeShiftR` bitsPerSubkey)
+                         | otherwise              = Invalid (INV2_misplaced_hash h0) p0
```

Comment:
I suspect the bug is in here. We seem to match up the hash with the sub-hash path in inverted order.